File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    
    graph=[[0]*(n+1) for _ in range(n+1)]
    for x,y in wires:
        graph[x][y]=1
        graph[y][x]=1
    
    
    #연결된 bfs로 몇개 연결 되었는지 확인
    def bfs(start):
        visited=[0]*(n+1)
        cnt=1
        que=deque([start])
        visited[start]=1
        while que:
            v=que.popleft() #어디서 끊기는지
            # print(v)
            for i in range(1,n+1):
                if graph[v][i]==1 and visited[i]==0:
                    que.append(i)
                    visited[i]=1
                    cnt+=1
        return cnt
            
    # 완전 탐색으로 전력망 끊고 붙이면서 제일 적은거 확인해보기
    # for a,b in wires:
    # graph[1][3]=0
    # graph[3][1]=0
    # print(bfs(1))
    # print(bfs(3))
    res=[]
    for a,b in wires:
        graph[a][b]=0
        graph[b][a]=0
        
        res.append(abs(bfs(a)- bfs(b)))
        
        graph[a][b]=1
        graph[b][a]=1
    
    answer = min(res)
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    static List<Integer> [] graph;
    static boolean [] visited;
    static int answer = Integer.MAX_VALUE;
    static int cnt;
    public int solution(int n, int[][] wires) {
  
        for(int k=0; k<wires.length; k++) {
            visited = new boolean[101];
            graph = new ArrayList [101];
            for(int j=0; j<101; j++) {
                graph[j] = new ArrayList<>();
            }
            for(int i=0; i<n-1; i++) {
                if(k!=i) {
                    graph[wires[i][0]].add(wires[i][1]);
                    graph[wires[i][1]].add(wires[i][0]);
                }
            }
            
                //System.out.println(Arrays.toString(graph));
                cnt=1;
                dfs(wires[k][0]);
                int a = cnt;
                cnt=1;
                dfs(wires[k][1]);
                int b = cnt;
                //System.out.println(a+" " +b);
                answer = Math.min(answer, Math.abs(a-b));
        }


        return answer;
    }
    //w에 따라 몇개 송신탑 있는지 세는 함수
    static void dfs(int w) {
        visited[w] = true;
        for(int i=0; i<graph[w].size(); i++) {
            int num = graph[w].get(i);
            if(!visited[num]) {
                visited[num] = true;
                cnt++;
                dfs(num);
            }
        }
    }
}